sdk/identity/identity/src/client/errors.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+/**
+ * See the official documentation for more details:
+ *
+ * https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-protocols-oauth-code#error-response-1
+ */
+export interface ErrorResponse {
+  error: string;
+  error_description: string;
+  error_codes?: number[];
+  timestamp?: string;
+  trace_id?: string;
+  correlation_id?: string;
+}
+
+export class AuthenticationError extends Error {
+  public readonly statusCode: number;
+  public readonly errorResponse: ErrorResponse;
+
+  constructor(statusCode: number, errorBody: string | undefined | null) {
+    let errorResponse = {
+      error: "unknown",
+      error_description: "An unknown error occurred and no additional details are available."
+    };
+
+    if (errorBody) {
+      try {
+        // Most error responses will contain JSON-formatted error details
+        // in the response body
+        errorResponse = JSON.parse(errorBody);
+      } catch (e) {
+        if (statusCode === 400) {
+          errorResponse = {
+            error: "authority_not_found",
+            error_description: "The specified authority URL was not found."
+          };
+        } else {
+          errorResponse = {
+            error: "unknown_error",
+            error_description: `An unknown error has occurred. Response body:\n\n${errorBody}`
+          };
+        }
+      }
+    } else {
+      errorResponse = {
+        error: "unknown_error",
+        error_description: "An unknown error occurred and no additional details are available."
+      };
+    }
+
+    super(
+      `An error was returned while authenticating to Azure Active Directory (status code ${statusCode}).\n\nMore details:\n\n${JSON.stringify(
+        errorResponse,
+        null,
+        "  "
+      )}`
+    );
+    this.statusCode = statusCode;
+    this.errorResponse = errorResponse;
+
+    // Restore the object prototype since Error's constructor futzes with it
+    Object.setPrototypeOf(this, new.target.prototype);
```

Comment:
What do you mean here? Can you show an example of the futzing?

Reply:
When you subclass `Error`, it hijacks the object prototype such that the resulting object looks like `Error` instead of the subclass itself.  There's a good example of it in the TypeScript 2.2 release notes:

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget

Reply:
Removed the hack in https://github.com/Azure/azure-sdk-for-js/pull/3826/commits/894d0b22479237262f651e7a61429b3f63af9e9d